pcdsdevices/ccm.py

```diff
@@ -773,6 +773,178 @@ def inverse(self, real_pos: namedtuple) -> namedtuple:
         return self.PseudoPosition(energy=energy)
 
 
+class CCMEnergyWithSelfSeed(CCMEnergy):
+    """
+    CCM energy motor with the self-seeding scheme.
+
+    Moves the alio to the requested energy, while also writing to the
+    BeamEnergyRequest PV so that ACR can set the self-seed energy accordingly.
+
+    The goal is to optimize the CCM energy around the self-seed energy. The
+    process is as follow:
+        1. Pass a setpoint energy to both the CCM and self-seeding.
+        2. Wait for self-seeding to stabilize.
+        3. Optimize the CCM throughput at the new energy.
+    There can be several implementations of the optimization step. Look at
+    the specific implementation methods for detailed information.
+
+    Parameters
+    ----------
+    prefix : str
+        The PV prefix of the Alio motor, e.g. XPP:MON:MPZ:07A
+    ipms_prefix : str
+        The EventBuilder PV containing the IPMs information
+    ipm_idx : dict
+        Indices in the EventBuilder array PV that contains the upstream
+        and downstream IPM. Format: {'upstream': 0, 'downstream': 1}
+    hutch : str, optional
+        The hutch we're in. This informs us as to which vernier
+        PVs to write to. If omitted, we can guess this from the
+        prefix.
+    """
+    self_seed = FCpt(BeamEnergyRequest, '{hutch}', kind='normal',
+                     doc='Requests ACR to move the self-seed energy.')
+
+    # IPMs intensities from the EventBuilder IOC. We need beam-synchronous IPM
+    # information, hence the use of the EventBuilder PV.
+    # See https://confluence.slac.stanford.edu/display/PCDS/Notes+on+the+Timetool+IOC
+    ipms = UCpt(EpicsSignalRO)
+
+    # These are duplicate warnings with main energy motor
+    _enable_warn_constants: bool = False
+    hutch: str
+
+    tab_component_names = True
+
+    def __init__(
+        self,
+        prefix: str,
+        ipms_prefix: str = None,
+        ipm_idx: dict = {'upstream': 0, 'downstream': 1},
+        hutch: typing.Optional[str] = None,
+        **kwargs
+    ):
+        # Put some effort into filling this automatically
+        # CCM exists only in two hutches
+        if hutch is not None:
+            self.hutch = hutch
+        elif 'XPP' in prefix:
+            self.hutch = 'XPP'
+        elif 'XCS' in prefix:
+            self.hutch = 'XCS'
+        else:
+            self.hutch = 'TST'
+
+        prefixes = {'ipms_prefix': ipms_prefix}
+        UCpt.collect_prefixes(self, prefixes)
+        super().__init__(prefix, **kwargs)
+
+        self.ipm_idx = ipm_idx
+
+        self.method = 'monotonic'
```

Comment:
Just spitballing here.  Maybe it'd be easier to have something like:
``` python
self.opt_args = {
    'method_name_1': {'method1_arg1': 1, 'method1_arg2': 2},
    'method_name_2': {'method2_arg1': 1, 'method2_arg2': 2},
}
```
Such that you can access the optimization args by method name.  That just gives you a place to put different args as you add more methods

Reply:
Implemented it, let me know what you think